crates/provider-sdk/src/provider.rs

```diff
@@ -196,109 +212,134 @@ impl ProviderConnection {
-                                    }
-                                    let current = tracing::Span::current();
-                                    current.record("operation", &tracing::field::display(&inv.operation));
-                                    current.record("lattice_id", &tracing::field::display(&lattice));
-                                    current.record("actor_id", &tracing::field::display(&inv.origin.public_key));
-                                    current.record("inv_id", &tracing::field::display(&inv.id));
-                                    current.record("host_id", &tracing::field::display(&inv.host_id));
-                                    current.record("provider_id", &tracing::field::display(&inv.target.public_key));
-                                    current.record("contract_id", &tracing::field::display(&inv.target.contract_id));
-                                    current.record("link_name", &tracing::field::display(&inv.target.link_name));
-                                    current.record("payload_size", &tracing::field::display(&inv.content_length));
-                                    let inv_id = inv.id.clone();
-                                    let inv_operation = inv.operation.clone();
-                                    let resp = match this.handle_rpc(provider.clone(), inv).in_current_span().await {
-                                        Err(err) => {
-                                            error!(%err, operation = %inv_operation, "Invocation failed");
-                                            InvocationResponse{
-                                                invocation_id: inv_id,
-                                                error: Some(format!("Error when handling invocation: {err}")),
-                                                ..Default::default()
+        let mut subs = vec![
+            self.rpc_client
+                .client()
+                .queue_subscribe(
+                    self.provider_rpc_topic(),
+                    RPC_SUBSCRIPTION_QUEUE_GROUP.to_string(),
+                )
+                .await?,
+        ];
+
+        // If we have a WIT resolve present, then this is a WIT-driven provider, and we can generate wprc
+        // topics on which we should listen to for RPCs
+        if let Some(resolve) = self.wit_resolve.as_ref() {
+            for topic in generate_incoming_rpcs_topics(&lattice, WRPC_VERSION, resolve)? {
+                subs.push(
+                    self.rpc_client
+                        .client()
+                        .queue_subscribe(topic, RPC_SUBSCRIPTION_QUEUE_GROUP.to_string())
+                        .await?,
+                )
+            }
+        }
+
+        // For every subscription, spawn a task to listen to incoming RPCs received,
+        // keeping track of the handles generated
+        let mut handles = Vec::new();
+        for mut sub in subs {
```

Comment:
Super nit: Couldn't we just combine all the streams using `futures::stream::SelectAll` and then spawn a single task? This might not matter with wRPC since the logic there might be different (since we can get all the info about the operation name from the topic).

Definitely don't worry about this yet until you have a bit more info on the wRPC stuff, was just calling that out

Reply:
We definitely could -- depending on task placement we might not want this all concentrated in one thread though -- so this is probably an improvement to make in the future if needed IMO